### PR TITLE
FgFlex: concat attention outputs w/ correct task convolution outputs

### PR DESCRIPTION
**Pull request by pmhalvor**
Realized that just adding tensors probably is erasing some of the information learned.
By concatenating all the information learned for a given task in a given stack, then re-encoding to the next stack's shape, we ensure all bits of information are maintained until the end of each stack. 

Another way of doing this would have been to expand the cnn_dim for each stack. 
While that probably would have an affect on results, I'll stick to re-encoding rn since others have used this before. 
This alternative can be a part of future work.

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -362,8 +362,8 @@ def __init__(
         self.bert.requires_grad = self.finetune
         self.bert_dropout = torch.nn.Dropout(self.dropout).to(torch.device(self.device)) 
         
-        # architecture specific components
-        self.components = torch.nn.ModuleDict(self.init_components(self.subtasks))  # returns dict of task-specific output layers
+        # architecture specific components: init_comp returns dict of task-specific output layers
+        self.components = torch.nn.ModuleDict(self.init_components(self.subtasks)).to(torch.device(self.device))
```
Comment (pmhalvor) [inline src/model.py:366]: Can never be too safe when working on GPUs

File: src/model.py [hunk 2]
```diff
@@ -1127,16 +1127,14 @@ def init_components(self, subtasks):
             if lstm_tasks[task]:
                 print(f"Adding lstm layer for {task}")
                 # add lstm to subtask's components
-                components[task].update({
-                    "lstm": torch.nn.LSTM(
+                components[task]["lstm"] = torch.nn.LSTM(
```
Comment (pmhalvor) [inline src/model.py:1130]: This PR also creates new dict objects more Pythonic (less scopy)

File: src/model.py [hunk 3]
```diff
@@ -1819,7 +1807,7 @@ def forward(self, batch):
             target_cnn = self.components["target"]["cnn"](target_inputs[-1])
             expression_cnn = self.components["expression"]["cnn"](expression_inputs[-1])
 
-            #### Relation R1  
+            ### Relation: R1
```
Comment (pmhalvor) [inline src/model.py:1810]: Also added some comments for the RACL relations

File: src/model.py [hunk 4]
```diff
@@ -2002,79 +1969,62 @@ def init_components(self, subtasks):
...
-                # in features are concats of embeddings, shared layers, and task cnns
-                "linear": self.linear_block(in_features=cnn_dim*3, out_features=3),
-
-                # map subtask information back to shared hidden state size
-                "re_encode": self.linear_block(
-                    in_features=cnn_dim*2,  # after attention
-                    out_features=cnn_dim
-                )
-            })
+            task: torch.nn.ModuleDict({})
```
Comment (pmhalvor) [inline src/model.py:1975]: Create empty ModuleDict, instead of `linear` and `re_encode`.
These were both refactored to where they are used 
(`linear` -> `linear_{stack}`; `re_encode` -> `stack-relation-first_task-re_encode`)

File: src/model.py [hunk 5]
```diff
@@ -2002,79 +1969,62 @@ def init_components(self, subtasks):
...
+                        components[task][f"cnn_{stack}"] = torch.nn.Sequential(*[
+                            self.cnn_block(cnn_dim*2, cnn_dim, kernel_size)
                             for layer in range(task_layers)
-                            ])}
-                        )
-                
+                        ])
+        
+                    ## Feedforward component: Expands per stack due to concatentation
+                    components[task][f"linear_{stack}"] = self.linear_block(in_features=(cnn_dim*2), out_features=3)
```
Comment (pmhalvor) [inline src/model.py:2008]: New stack-wise linears

File: src/model.py [hunk 6]
```diff
@@ -2002,79 +1969,62 @@ def init_components(self, subtasks):
...
 
+        else: 
+            for task in subtasks:
+                # final outputs for each task
+                # in_features are concats of embeddings, shared layers, and task cnns
+                components[task]["linear"] = self.linear_block(in_features=cnn_dim*2, out_features=3)
```
Comment (pmhalvor) [inline src/model.py:2014]: Single linear when stack = 0, also handled like this in `forward`

File: src/model.py [hunk 7]
```diff
@@ -2002,79 +1969,62 @@ def init_components(self, subtasks):
...
         ######################################
         # Shared relation components
         ######################################
 
-        # similar to IMN setup
-        # NOTE main difference: IMN uses split_cnn_block w/ kernel=3 & kernel=5
-        if split_cnn_kernels is not None:  # NOTE only split on first shared
+        ## Shared layers: similar to IMN setup
+        # default difference: IMN uses split_cnn_block w/ kernel=3 & kernel=5
+        if split_cnn_kernels is not None:  # NOTE only splitting on first shared
```
Comment (pmhalvor) [inline src/model.py:2021]: Only splitting on first shared cnn bc thats how its done in IMN.

Future work can look into implementing this for all shared, if that could possible give even better performance.

File: src/model.py [hunk 8]
```diff
@@ -2083,44 +2033,42 @@ def init_components(self, subtasks):
...
-        if stack_count > 0:  # when stack == 0 no attention should be made. No task-wise cnn either.. 
-            prev_task = None
-            for task in self.subtasks:
-                for relation in components["relations"]["stack_0"]:
-                    if relation.startswith(task) and (task != prev_task):
-                        components[task]["attn_linear"] = self.linear_block(cnn_dim*2, 3)
-                        prev_task = task
+        ## Relation layers: similar to RACL setup
+        relations = torch.nn.ModuleDict({})
+        attn_linears = torch.nn.ModuleDict({})
```
Comment (pmhalvor) [inline src/model.py:2049]: This isn't used, delete.

File: src/model.py [hunk 9]
```diff
@@ -2083,44 +2033,42 @@ def init_components(self, subtasks):
...
+        for stack in range(stack_count):
+            for rel in attention_relations:
+                relations[f"stack_{stack}"] = torch.nn.ModuleDict({
+                    f"{rel[0]}_at_{rel[1]}": torch.nn.ModuleDict({
+                        "attn": self.attn_block(cnn_dim),
+
+                        # new linear for each attn relation for first_task logits using attn information
+                        "linear": self.linear_block(cnn_dim*2, 3),
+                    })
+                })
```
Comment (pmhalvor) [inline src/model.py:2059]: Each relation now consists of an attention layer and a linear layer for producing task-wise logits

File: src/model.py [hunk 10]
```diff
@@ -2083,44 +2033,42 @@ def init_components(self, subtasks):
...
+                if rel[0] in self.subtasks + ["shared"] and "re_encode" not in components[rel[0]].keys():
+                    # only reach this one time per first_task=rel[0]
+                    count = sum([rel[0] == r[0] for r in attention_relations])
+                    # map subtask information back to shared hidden state size after (multiple) attention(s)
+                    components[rel[0]]["re_encode"] = self.linear_block(
+                        in_features=cnn_dim*(1 + count),  # 1 task output + number of relations with rel[0] as first task
+                        out_features=cnn_dim*2
+                    )
```
Comment (pmhalvor) [inline src/model.py:2067]: A lot going on here.

First off, the if-statement:
We want to store the outputs of all the attentions a given subtask is a part of, before updating the next task input. The if checks that `first_task` is a subtask, and that it doesn't have a re-encoding layer form before.

`count`:
We get the number of times the current `first_task` occurs as a first task in `attention_relations`. This way, we know how many extra attn outputs will be concatenated.

Linear `re_encode` layer:
The input features to this layer need to account for this task's cnn output, plus all the times this task is a `first_task` in an attention. This way, we use all the information from every attention block during re-encoding for th next stack.

File: src/model.py [hunk 11]
```diff
@@ -2153,7 +2101,7 @@ def forward(self, batch):
             attention_mask = attention_mask,
         ).last_hidden_state
         embeddings = self.bert_dropout(embeddings).permute(0, 2, 1)
-        sentence_output = embeddings.clone()
+        shared_output = embeddings.clone()
```
Comment (pmhalvor) [inline src/model.py:2104]: Changed name to better match naming in task-specific loops (for better human readability)

File: src/model.py [hunk 12]
```diff
@@ -2163,38 +2111,34 @@ def forward(self, batch):
...
-            sentence_output = shared[f"cnn_{i}"](sentence_output)
+            shared_output = shared[f"cnn_{i}"](shared_output)
 
-        # update word embeddings with shared features learned from this cnn layer
-        embeddings = torch.cat((embeddings, sentence_output), dim=1) # cat embedding dim
-
-        # only the information learned from shared cnn(s), no embeddings
-        initial_shared_features = sentence_output
+        # initial shared features consists of embedding plus shared cnn output
+        initial_shared_features = torch.cat((embeddings, shared_output), dim=1) # cat embedding dim
```
Comment (pmhalvor) [inline src/model.py:2128]: Now, our initial shared features contains out embeddings and shared cnn output.

File: src/model.py [hunk 13]
```diff
@@ -2163,38 +2111,34 @@ def forward(self, batch):
...
 
         ######################################
         # Stacked relation interactions
         ######################################
-        task_inputs = {  # TODO document
+        task_inputs = {  # initial shared features consists of embedding plus shared cnn output
```
Comment (pmhalvor) [inline src/model.py:2133]: Some comment documenting

File: src/model.py [hunk 14]
```diff
@@ -2211,25 +2155,33 @@ def forward(self, batch):
                         ],
                         dim=1
                     )
-                else:
+                elif len(self.components[task][f"cnn_{stack}"]) > 0:
+
```
Comment (pmhalvor) [inline src/model.py:2159]: Previous, we were relying on empty sequential outputs just being input. But now, since `cnn_dim` changes in first cnn block, we needed to handle when some tasks don't use cnns.
(This is the part that could have been implmented differently, but will now be saved for later work)

File: src/model.py [hunk 15]
```diff
@@ -2211,25 +2155,33 @@ def forward(self, batch):
                         ],
                         dim=1
                     )
-                else:
+                elif len(self.components[task][f"cnn_{stack}"]) > 0:
+                        
                     cnn_outputs[task] = self.components[task][f"cnn_{stack}"](task_output)
+                else: 
+                    cnn_outputs[task] = shared_output
```
Comment (pmhalvor) [inline src/model.py:2162]: Since no cnn output for this task, just reusing shared outputs.

File: src/model.py [hunk 16]
```diff
@@ -2211,25 +2155,33 @@ def forward(self, batch):
...
             cnn_outputs["all"] = sum([cnn_outputs[task] for task in self.subtasks])
 
             ######################################
             # Subtask relations mimic RACL setup
             ######################################
             prev_first = None
-            same_task = []
+            relation_inters = {}
             for relation in self.components["relations"][f"stack_{stack}"]:
+                relation_components = self.components["relations"][f"stack_{stack}"][relation]
```
Comment (pmhalvor) [inline src/model.py:2177]: For easier classing further downstream (keys are attn and linear)

File: src/model.py [hunk 17]
```diff
@@ -2259,35 +2211,34 @@ def forward(self, batch):
...
 
-                # store task inputs for next stack, considering multiple relations for first task
-                if (first_task == prev_first) or (first_task == "shared") or (prev_first is None):
-                    same_task.append(attn_inter)
+                # append attn output to list if first task present, else create new list
+                if first_task in relation_inters:
+                    relation_inters[first_task].append(relation_attn.permute(1,0,2))
                 else:
-                    same_task = [attn_inter]
+                    relation_inters[first_task] = [relation_attn.permute(1,0,2)]
```
Comment (pmhalvor) [inline src/model.py:2225]: Storing all attention outputs according to the first task

File: src/model.py [hunk 18]
```diff
@@ -2259,35 +2211,34 @@ def forward(self, batch):
...
 
-                if first_task in task_inputs:
-                    re_encoded = self.components[first_task]["re_encode"](torch.stack(same_task).mean(dim=0))
-                    task_inputs[first_task] = re_encoded.permute(0, 2, 1)
-                prev_first = first_task    
+            # concatenate all attention outputs using same first task w/ that task's conv outputs
+            for task in relation_inters:
+                if task in self.subtasks:
+                    all_task_inters = torch.cat([cnn_outputs[task].permute(0, 2, 1)] + relation_inters[task], dim=-1)
+                    task_inputs[task] = self.components[task]["re_encode"](all_task_inters).permute(0, 2, 1)  # batch, sequence, embedding
```
Comment (pmhalvor) [inline src/model.py:2231]: Re-encoding outputs from all attentions outputs and convolution output for these tasks.

File: src/model.py [hunk 19]
```diff
@@ -2319,7 +2270,7 @@ def transmission(self, logits, true):
 
         focus_scope = (  # strengthen signal from true logits
             gold_influence*true.bool().float()
-                + (1-gold_influence)*logits.argmax(-1)
+                + (1-gold_influence)*logits.argmax(-1).bool().float()
```
Comment (pmhalvor) [inline src/model.py:2273]: Don't count `B` as twice as important as `I`s

**Review by pmhalvor — COMMENTED** [review]
Comment pointing out and explaining major changes.

Important to note:
- Tasks involved as first tasks in an attention relation have their outputs concated w/ cnn output, then re-encoded for the next stacks task inputs. 
- Dictionary updates replaced by `=` key: value instantiation. 
- Transmission now does not count B as 2x I, but rather treats labels > 0 as 1 and 0 otherwise
- Task inputs are now both plain embeddings and shared cnn output (for first stack)
